packages/opencensus-core/src/trace/instrumentation/base-plugin.ts

```diff
@@ -44,26 +64,111 @@ export abstract class BasePlugin implements types.Plugin {
    * @param moduleExports nodejs module exports to set as context
    * @param tracer tracer relating to context
    * @param version module version description
+   * @param basedir module absolute path
    */
   // tslint:disable:no-any
-  protected setPluginContext(
-      moduleExports: any, tracer: modelTypes.Tracer, version: string) {
+  private setPluginContext(
+      moduleExports: any, tracer: modelTypes.Tracer, version: string,
+      basedir?: string) {
     this.moduleExports = moduleExports;
     this.tracer = tracer;
     this.version = version;
+    this.basedir = basedir;
+    this.logger = tracer.logger;
+    this.internalFilesExports = this.loadInternalFiles();
+  }
+
+
+  /**
+   * Method to apply the instrumentation patch.
+   *
+   * GoF Template Method Pattern - this is the invariant part of the pattern
+   *
+   * @param moduleExports nodejs module exports from the module to patch
+   * @param tracer a tracer instance
+   * @param version version of the current instaled module to patch
+   * @param basedir module absolute path
+   */
+  enablePluginPatch(
+      // tslint:disable:no-any
+      moduleExports: any, tracer: modelTypes.Tracer, version: string,
+      basedir: string) {
+    this.setPluginContext(moduleExports, tracer, version, basedir);
+    return this.applyPatch();
   }
 
 
-  // TODO: review this implementation
-  // From the perspective of an instrumentation module author,
-  // that applyUnpatch is abstract makes it seem like patching is optional,
-  // while unpatching is not. It should be the other way around
+  disablePluginPatch() {
+    this.applyUnpatch();
+  }
 
+  /**
+   * applyPatch() and applyUnpatch()
+   *
+   * GoF Template Method Pattern - this is the variant part of the pattern
+   * Each plugin should implement his own version.
```

Comment:
I would not consider this universal language. I did not know what this was referring to and had to look it up. It might be true that I am not the common case, but it is better not to assume something like this about the audience. After all, it is possible to understand the concept without knowing its name.

Feel free to document the design pattern at the top of the file, but I do not think it belongs here.

As another tidbit, we should not use gender pronouns in code, especially when referring to impersonal objects. Please use `its`.

Reply:
> ... document the design pattern at the top of the file ...

At the top, it loses context. I have improved the message.